docs: add north-south messaging (#887)
edgexfoundry/edgex-docs: docs_src/microservices/core/command/Ch-Command.md

## Configuration Properties

Please refer to the general [Common Configuration documentation](../../configuration/CommonConfiguration.md) for configuration properties common to all services.
Below are only the additional settings and sections that are not common to all EdgeX Services.

!!! edgey "EdgeX 2.3"

>     `RequireMessageBus` and `MessageQueue` configurations were added in EdgeX 2.3 for North-South Messaging

Reviewer: Missed one
```suggestion
    `RequireMessageBus` and `MessageQueue` configurations were added in EdgeX 2.3 for Commands via Messaging
```

> === "RequireMessageBus"

Reviewer 2: The Device Services (device-sdk-c, at least) have "UseMessageBus" to determine whether to publish Event updates on the Message Bus. 
That's a confusingly similar name to "RequireMessageBus", with a different meaning. Probably need to add something about "UseMessageBus" too, to keep them apart in understanding.

Author: @lenny-intel do you think we need to update this config to `UseMessageBus` ?

Reviewer 2: OK, makes sense.

=== "MessageQueue.Internal.Topics"
    |Property|Default Value|Description|
    |---|---|---|
    |||Key-value mappings allow for publication and subscription to the internal message bus |
    | DeviceRequestTopicPrefix | `edgex/device/command/request` | For publishing requests to the device service. `<device-service>/<device-name>/<command-name>/<method>` will be added to this publish topic prefix |
    | DeviceResponseTopic | `edgex/device/command/response/#` | For subscribing to device service responses |
    | CommandRequestTopic | `edgex/core/command/request/#` | For subscribing to internal command requests |
    | CommandResponseTopicPrefix | `edgex/core/command/response` | For publishing responses back to internal service. `/<device-name>/<command-name>/<method>` will be added to this publish topic prefix |
    | QueryRequestTopic | `edgex/core/commandquery/request/#` | For subscribing to internal command query requests |
    | QueryResponseTopic | `edgex/core/commandquery/response` | For publishing command query responses back to internal service |
=== "MessageQueue.Internal.Optional"
    |Property|Default Value|Description|
    |---|---|---|

>     |||Configuration and connection parameters for use with MQTT message bus - in place of Redis|

Reviewer 2: ```suggestion
    |||Configuration and connection parameters used when Type is 'mqtt', instead of 'redis'|
```

## North-South Messaging

!!! edgey "Edgex 2.3"
    North-South Messaging is new in EdgeX 2.3

### Introduction
Previously, communications from a 3rd party system (enterprise application, cloud application, etc.) to EdgeX in order to acuate a device or get the latest information from a sensor was only accomplished via REST.
The 3rd party system makes a REST call of the command service which then relays a request to a device service also using REST.
There was no built-in means to make a message-based request of EdgeX or the devices/sensors it manages.

From Levski release, core command service adds support for an external MQTT connection (in the same manner that app services provide an external MQTT connection),
which will allow it to act as a bridge between the internal message bus (implemented via either MQTT or Redis Pub/Sub) and external MQTT message bus.

#### Core Command as Message Bus Bridge

The Core Command service will serve as the EdgeX entry point for external, north-to-south message bus requests to the south side.

![image](../../../design/adr/command-msg.png)

> 3rd party systems should not be granted access to the EdgeX internal message bus. Therefore, in order to implement north to south communications via message bus (specifically MQTT), the command service needs to take messages from the 3rd party or external MQTT topics and pass them internally onto the EdgeX internal message bus where they can eventually be routed to the device services and then on to the devices/sensors (southside).

Author: This seems odd to me.
I removed the north to south for now.

> 3rd party systems should not be granted access to the EdgeX internal message bus. Therefore, in order to implement north to south communications via message bus (specifically MQTT), the command service needs to take messages from the 3rd party or external MQTT topics and pass them internally onto the EdgeX internal message bus where they can eventually be routed to the device services and then on to the devices/sensors (southside).

Reviewer: Because my change wasn't complete? ;-)

Should have been `implement commands via message bus`

}
```

The messages for formatted requests and responses are sharing a common base structure.
The outermost JSON object represents the message envelope, which is used to convey metadata about request/response including `ApiVersion`, `RequestID`, `CorrelationID`...etc.

The `Payload` field contains the base64-encoded response body.  
The `ErrorCode` field provides the indication of error.
The `ErrorCode` will be 0 (no error) or 1 (indicating error) as the two enums for error conditions.
When there is an error (with `ErrorCode` set to 1), then the payload contains a message string indicating more information about the error.
When there is no error (errorCode 0) then there is no message string in the payload.

### Command Query

Core Command service subscribes to the `QueryRequestTopic` and publishes the response to `QueryResponseTopic` defined in the configuration file.
After receiving the request, Core Command service will try to parse the `<device-name>` from request topic level.
The 3rd party system or application must publish command query requests messages and subscribe to responses from the same topics.
Below is the default topic naming used by Core Command:

> - Subscribing command query request topic: `edgex/commandquery/request/<device-name>`

Reviewer: ```suggestion
- Subscribing command query request topic: `edgex/commandquery/request/#`
```

```json
{
  "ApiVersion": "v2",
  "ContentType": "application/json",
  "CorrelationID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
  "RequestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
  "QueryParams": {
    "offset": "0",
    "limit": "5"
  }
}
```

2. Receive query response message from external MQTT broker on topic `edgex/commandquery/response`:
```json
{
  "ApiVersion":"v2",
  "ContentType":"application/json",
  "CorrelationID":"14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
  "RequestID":"e6e8a2f4-eb14-4649-9e2b-175247911369",
  "ErrorCode":0,
  "Payload":"..."
}
```

### Command Request

Core Command service subscribes to the `CommandRequestTopic` defined in the configuration file.
After receiving the request, Core Command service will try to parse `<device-name>` `<command-name>` and `<method>` from request topic level,
and send the response back with `<device-name>`, `<command-name>` and `<method>` appended to `CommandResponseTopicPrefix` defined in the configuration file.
The 3rd party system or application must publish command requests messages and subscribe to responses from the same topics.
Below is the default topic naming used by Core Command:

> - Subscribing command request topic: `edgex/command/request/<device-name>/<command-name>/<method>`

Reviewer: ```suggestion
- Subscribing command request topic: `edgex/command/request/#`
```
